### PR TITLE
internal/timeout: extract package for handling timeouts

### PR DESCRIPTION
**Pull request by skriss**
WIP, let me know if this makes sense. I got sick of keeping track of what -1, 0, "", and "infinity" meant between Contour and Envoy so added a proper type for it.

If this makes sense, then we can use the same package for processing the config file timeout values so the syntax is consistent everywhere.

Updates #2697

### CODE REVIEW

**Comment by jpeach** [comment]
@skriss I think this makes a lot of sense and is a great improvement. I'd like to do a full review next week.

File: internal/timeout/timeout.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+// Setting describes a timeout setting that can be exactly one of:
+// disable the timeout entirely, use the default, or use a specific
+// value. The zero value is a Setting representing "use the default".
+type Setting struct {
+	val     time.Duration
+	disable bool
+}
+
+// Disable returns whether the timeout should be disabled entirely.
+func (s Setting) Disable() bool {
```
[the remark this answers is not in the document]
Comment (skriss) [reply internal/timeout/timeout.go]: yep that's a clearer name, updated.

File: internal/envoy/route.go [hunk 2]
```diff
@@ -135,33 +135,33 @@ func responseTimeout(r *dag.Route) *duration.Duration {
 	if r.TimeoutPolicy == nil {
 		return nil
 	}
-	return timeout(r.TimeoutPolicy.ResponseTimeout)
+	return envoyTimeout(r.TimeoutPolicy.ResponseTimeout)
 }
 
 func idleTimeout(r *dag.Route) *duration.Duration {
```
Comment (jpeach) [inline internal/envoy/route.go:141]: You can clean this up further by making the `dag.Route.TimeoutPolicy` a non-pointer member that defaults all its fields to `setting.DefaultSetting()`. Then you would  not need these wrappers just to fetch specific fields, you could just do `envoyTimeout(r.TimeoutPolicy.ResponseTimeout)` directly in the caller.

File: internal/dag/policy.go [hunk 3]
```diff
@@ -147,8 +158,8 @@ func timeoutPolicy(tp *projcontour.TimeoutPolicy) *TimeoutPolicy {
 		return nil
 	}
 	return &TimeoutPolicy{
-		ResponseTimeout: annotation.ParseTimeout(tp.Response),
-		IdleTimeout:     annotation.ParseTimeout(tp.Idle),
+		ResponseTimeout: timeout.Parse(tp.Response),
+		IdleTimeout:     timeout.Parse(tp.Idle),
```
Comment (jpeach) [inline internal/dag/policy.go:162]: Now that you have explicit defaults, you can return a default policy if `tp` is nil:
```
if tp == nil {
        return &TimeoutPolicy {
                ResponseTimeout: timeout.DefaultSetting(),
                IdleTimeout: timeout.DefaultSetting(),
        }
}
```

This lets you clean up helpers, per comment below.

File: internal/timeout/timeout_test.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	}
+}
+
+func TestWithDuration(t *testing.T) {
+	s := WithDuration(10 * time.Second)
+	want := 10 * time.Second
+	got := s.Val()
+	if want != got {
+		t.Errorf("Wanted %v, got %v", want, got)
+	}
```
[the remark this answers is not in the document]
Comment (skriss) [reply internal/timeout/timeout_test.go]: I had to make a change to the `assert` package to remove an import cycle that was preventing me from using it in the timeout tests, but it's now simplified.

File: internal/timeout/timeout.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+// UseDefault is a Setting representing "use the default".
+var UseDefault = Setting{}
+
+// Parse parses string representations of timeout settings that we pass
+// in various places in a standard way:
+//	- an empty string means "use the default".
+//	- any valid representation of "0" means "use the default".
+//	- a valid Go duration string is used as the specific timeout value.
+//	- any other input means "disable the timeout".
+func Parse(timeout string) Setting {
```
[the remark this answers is not in the document]
Comment (skriss) [reply internal/timeout/timeout.go:63]: I'll tackle that in a follow-up.

File: internal/envoy/route.go [hunk 6]
```diff
@@ -180,8 +180,8 @@ func retryPolicy(r *dag.Route) *envoy_api_v2_route.RetryPolicy {
 	if r.RetryPolicy.NumRetries > 0 {
 		rp.NumRetries = protobuf.UInt32(r.RetryPolicy.NumRetries)
 	}
-	if r.RetryPolicy.PerTryTimeout > 0 {
-		rp.PerTryTimeout = protobuf.Duration(r.RetryPolicy.PerTryTimeout)
+	if val := r.RetryPolicy.PerTryTimeout.Val(); val > 0 {
+		rp.PerTryTimeout = protobuf.Duration(val)
```
[the remark this answers is not in the document]
Comment (skriss) [reply internal/envoy/route.go]: ended up switching it to use the `envoyTimeout` helper.

**Comment by skriss** [comment]
Rebased to address merge conflicts.

**Pull request merged**